src/js/components/JSONEditor.js

```diff
@@ -0,0 +1,504 @@
+  /**
+   * `onFocus` handler for the AceEditor
+   *
+   * @param {DOMEvent} event - The event object
+   */
+  handleFocus(event) {
+    this.props.onFocus(event, this.jsonValue);
+    this.isFocused = true;
+  }
+
+  /**
+   * Apply pending state updates to the ACE editor instance.
+   *
+   * This function operates purely on the `aceEditor` instance, created by the
+   * `AceEditor` composed and exposed through the `onLoad` event. Since the
+   * `AceEditor` component does a pretty bad job updating only what's needed,
+   * we had to use this approach and avoid re-rendering the component as much
+   * as possible.
+   */
+  updateEditorState() {
+    if (!this.aceEditor) {
+      return;
+    }
+
+    // Prepare markers
+    let markers = [];
+
+    // Extract syntax errors, or other errors that refer to line
+    if (this.jsonError) {
+
+      // Strip out the 'at line xxx' message, and keep track of that line
+      let errorLine = 0;
+      let errorMsg = this.jsonError.replace(/at line ([\d:]+)/g,
+        function (m, line) {
+          errorLine = parseInt(line.split(':')[0]);
+          return '';
+        }
+      );
+
+      // Push the error marker
+      markers.push({
+        row: errorLine,
+        text: errorMsg,
+        type: 'error'
+      });
+
+    } else {
+
+      // Append annotation markers from external errors
+      markers = this.externalErrors.reduce((memo, error) => {
+        let errorPath = error.path.join('.');
+        let token = this.jsonMeta.find(function (token) {
+          return token.path.join('.') === errorPath;
+        });
+
+        if (!token) {
+          return memo;
+        }
+
+        memo.push({
+          row: token.line-1,
+          text: error.message,
+          type: 'error'
+        });
+
+        return memo;
+      }, markers);
+
+    }
+
+    // Merge current annotations with the annotations we are going to show
+    this.aceEditor.getSession().setAnnotations(markers);
+  }
+
+  /**
+   * This function assigns the local state of the JSON text, outside the React
+   * state, since this does not nesseceraly trigger an update.
+   *
+   * This is useful in order to have the latest available metainformation when
+   * the user changes the text (on the `handleChange` handler), but we don't
+   * want to re-do the processing when this will eventually trigger an update
+   * to the `value` property with the same data.
+   *
+   * @param {String} jsonText - The JSON buffer to extract metadata from
+   */
+  updateLocalJsonState(jsonText) {
+    // Do not perform heavyweight calculations, such as `getObjectInformation`
+    // if the json text hasn't really changed.
+    if (this.jsonText === jsonText) {
+      return;
+    }
+
+    // Reset some properties
+    this.jsonText = jsonText;
+    this.jsonError = null;
+
+    // Try to parse and extract metadata
+    try {
+      this.jsonValue = JSON.parse(jsonText);
+      this.jsonMeta = JSONUtil.getObjectInformation(jsonText).reverse();
```

Comment:
Hmm.. good catch.  This is a leftover from a previous implementation and I guess we can remove it completely. For the record:

The `getObjectInformation` extracts the tokens in the order they appear. Previously, I was tracking objects based on cursor position and this was problematic in the case of nested objects. For example, given the string `{a: {b: 'v'}}` it will return first the `{b: 'v'}` token and then the `v` token, meaning that when searching, I was getting the wrong token.